src/storage/mvcc/txn.rs

```diff
@@ -199,14 +178,45 @@ impl<S: Snapshot> MvccTxn<S> {
         Ok(())
     }
 
+    // Pessimistic transactions only acquire pessimistic locks on row keys.
+    // The corrsponding index keys are not locked until pessimistic prewrite.
+    // It's possible that lock conflict occours on them, but the isolation is
+    // guaranteed by pessimistic locks on row keys, so let TiDB resolves these
+    // locks immediately.
+    fn handle_non_pessimistic_lock_conflict(
+        &self,
+        key: Key,
+        for_update_ts: u64,
+        lock: Lock,
+    ) -> Result<()> {
+        // Optimistic lock's for_update_ts is its start_ts.
+        let lock_for_update_ts = if lock.for_update_ts > 0 {
+            lock.for_update_ts
+        } else {
+            lock.ts
+        };
+        if for_update_ts > lock_for_update_ts {
+            Err(Error::KeyIsLocked {
+                key: key.into_raw()?,
+                primary: lock.primary,
+                ts: lock.ts,
+                // Set ttl to 0 so TiDB will resolve lock immediately.
+                ttl: 0,
```

Comment:
what if this lock is a normal primary key, and the corresponding txn is not time out？